nexus/src/authn/silos.rs

```diff
@@ -205,4 +212,197 @@ impl SamlIdentityProvider {
             Ok(None)
         }
     }
+
+    pub fn authenticated_subject(
+        &self,
+        body_bytes: &str,
+        max_issue_delay: Option<chrono::Duration>,
+    ) -> Result<(AuthenticatedSubject, Option<String>), HttpError> {
+        // Given a post body, decode that as a SAMLResponse. With POST binding,
+        // the SAMLResponse will be POSTed to Nexus as a base64 encoded XML
+        // file, with optional relay state being whatever Nexus sent as part of
+        // the SAMLRequest.
+        let saml_post: SamlLoginPost = serde_urlencoded::from_str(body_bytes)
+            .map_err(|e| {
+            HttpError::for_bad_request(
+                None,
+                format!("urlencoded::from_str failed! {}", e),
```

Comment:
The various errors in this function don't seem oriented towards the users that are likely to see them.  It seems worth making these more useful for whoever's debugging why this operation failed.  (I'm imagining this will either be an Oxide support person or maybe someone at the customer site who's trying to figure out why the rack is rejecting the requests from the IdP?  Or maybe even an end user at the customer site?)